### PR TITLE
feat: catalog source polling for updated registry image

### PR DESCRIPTION
**Pull request by exdx**
**Description of the change:**
Branch consolidates implementation of catalog polling feature that continues from #1109. 

**Motivation for the change:**
New feature request. Part of https://github.com/openshift/enhancements/blob/master/enhancements/olm/operator-registry.md. 

**Reviewer Checklist**
- [x] Implementation matches the proposed design, or proposal is updated to match implementation
- [x] Sufficient unit test coverage 
- [x] Sufficient end-to-end test coverage
- [x] Docs updated or added to `/docs` 
- [x] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/controller/registry/reconciler/grpc.go [hunk 1]
```diff
@@ -16,6 +17,8 @@ import (
 	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/ownerutil"
 )
 
+const CatalogSourceUpdateKey = "olm.catalogSource.update"
```
Comment (njhale) [inline pkg/controller/registry/reconciler/grpc.go]: I think we should start sticking to the [conventions laid out in the k8s docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) for label keys. WDYT about something like `catalogsource.operators.coreos.com/update`?

File: pkg/controller/registry/reconciler/grpc.go [hunk 2]
```diff
@@ -168,6 +248,58 @@ func (c *GrpcRegistryReconciler) ensureService(source grpcCatalogSourceDecorator
...
+func updatePodByDigest(updatePod *corev1.Pod, servingPods []*corev1.Pod) bool {
+	newCatSrcImage := getPodImageID(updatePod)
+
+	if len(servingPods) > 0 {
+		for _, p := range servingPods {
+			if newCatSrcImage != getPodImageID(p) {
+				return true
+			}
+		}
+	}
```
Comment (njhale) [inline pkg/controller/registry/reconciler/grpc.go]: The length check around the for loop can be removed.

File: pkg/controller/registry/reconciler/grpc.go [hunk 3]
```diff
@@ -168,6 +248,58 @@ func (c *GrpcRegistryReconciler) ensureService(source grpcCatalogSourceDecorator
...
+	return false
+}
+
+func getPodImageID(pod *corev1.Pod) string {
+	if pod.Status.ContainerStatuses != nil && len(pod.Status.ContainerStatuses) > 0 {
+		return pod.Status.ContainerStatuses[0].ImageID
+	} else {
+		return ""
+	}
+}
```
Comment (njhale) [inline pkg/controller/registry/reconciler/grpc.go]: This `return` statement doesn't need to be in an `else`.

File: pkg/controller/registry/reconciler/grpc.go [hunk 4]
```diff
@@ -168,6 +248,58 @@ func (c *GrpcRegistryReconciler) ensureService(source grpcCatalogSourceDecorator
...
+
+func (c *GrpcRegistryReconciler) removePods(pods []*corev1.Pod, namespace string) error {
+	if len(pods) > 0 {
+		for _, p := range pods {
+			err := c.OpClient.KubernetesInterface().CoreV1().Pods(namespace).Delete(p.GetName(), metav1.NewDeleteOptions(0))
+			if err != nil {
+				return errors.Wrapf(err, "error deleting pod: %s", p.GetName())
+			}
+		}
+	}
```
Comment (njhale) [inline pkg/controller/registry/reconciler/grpc.go]: The length check can be removed.

File: pkg/api/apis/operators/v1alpha1/catalogsource_types.go [hunk 5]
```diff
@@ -66,6 +73,10 @@ type CatalogSourceSpec struct {
 	Icon        Icon   `json:"icon,omitempty"`
 }
 
+type Poll struct {
+	Interval metav1.Duration `json:"interval,omitempty"`
+}
+
```
[the remark this answers is not in the document]
Comment (exdx) [reply pkg/api/apis/operators/v1alpha1/catalogsource_types.go]: This is something we discussed with @ecordell after changing it once to your suggestion ourselves - the original design of this feature has the implementation with `Poll` and then an `Interval` subfield so we decided to stick to that. Happy to look at it further.

File: deploy/chart/templates/0000_50_olm_06-catalogsource.crd.yaml [hunk 6]
```diff
@@ -71,6 +71,16 @@ spec:
...
 
+            poll:
+              type: object
+              description: Poll represents the duration between successive checks of the catalog to obtain the latest version of the catalogsource
+              required:
+              - interval
+              properties:
+                interval:
+                  type: string
+                  description: The actual time interval between checks. Gets coerced into a time.Duration type after unmarshaling.
```
[the remark this answers is not in the document]
Comment (exdx) [reply deploy/chart/templates/0000_50_olm_06-catalogsource.crd.yaml]: Good point

File: pkg/controller/registry/reconciler/grpc.go (hunk 1 above)
Comment (exdx) [reply pkg/controller/registry/reconciler/grpc.go]: That sounds good to me

File: pkg/api/apis/operators/v1alpha1/catalogsource_types.go [hunk 7]
```diff
@@ -92,6 +103,9 @@ type CatalogSourceStatus struct {
 	// +optional
 	Reason ConditionReason `json:"reason,omitempty"`
 
+	// The last time catalogsource has been checked to ensure the image is up-to-date
```
[the remark this answers is not in the document]
Comment (exdx) [reply pkg/api/apis/operators/v1alpha1/catalogsource_types.go]: I changed it to `LatestImageRegistryPoll` hopefully more descriptive

File: Makefile [hunk 8]
```diff
@@ -22,7 +22,7 @@ IMAGE_TAG ?= "dev"
 SPECIFIC_UNIT_TEST := $(if $(TEST),-run $(TEST),)
 LOCAL_NAMESPACE := "olm"
 export GO111MODULE=on
-CONTROLLER_GEN := go run ./vendor/sigs.k8s.io/controller-tools/cmd/controller-gen
+CONTROLLER_GEN := go run -mod=vendor ./vendor/sigs.k8s.io/controller-tools/cmd/controller-gen
 YQ_INTERNAL := go run ./vendor/github.com/mikefarah/yq/v2/
```
Comment (ecordell) [inline Makefile]: isn't -mod-vendor needed here too?

File: Makefile [hunk 9]
```diff
@@ -158,9 +158,12 @@ codegen:
 	go run vendor/k8s.io/kube-openapi/cmd/openapi-gen/openapi-gen.go --logtostderr -i ./vendor/k8s.io/apimachinery/pkg/runtime,./vendor/k8s.io/apimachinery/pkg/apis/meta/v1,./vendor/k8s.io/apimachinery/pkg/version,./pkg/package-server/apis/operators/v1,./pkg/package-server/apis/apps/v1alpha1,./pkg/api/apis/operators/v1alpha1,./pkg/lib/version -p $(PKG)/pkg/package-server/apis/openapi -O zz_generated.openapi -h boilerplate.go.txt -r /dev/null
 	$(CODEGEN_INTERNAL) deepcopy,conversion,client,lister,informer $(PKG)/pkg/api/client $(PKG)/pkg/api/apis $(PKG)/pkg/api/apis "operators:v1alpha1,v1"
 	$(CODEGEN_INTERNAL) all $(PKG)/pkg/package-server/client $(PKG)/pkg/package-server/apis $(PKG)/pkg/package-server/apis "operators:v1 apps:v1alpha1"
+
+manifestgen:
```
Comment (ecordell) [inline Makefile]: shouldn't this be called by `codegen`? (or a prereq, whatever it's called in make)

File: pkg/controller/registry/reconciler/grpc.go [hunk 10]
```diff
@@ -136,22 +160,78 @@ func (c *GrpcRegistryReconciler) EnsureRegistryServer(catalogSource *v1alpha1.Ca
...
+			updatePod.Labels[CatalogSourceLabelKey] = source.GetName()
+			updatePod.Labels[CatalogSourceUpdateKey] = ""
+
+			// Update the update pod to promote it to serving pod
+			_, err := c.OpClient.KubernetesInterface().CoreV1().Pods(source.GetNamespace()).Update(updatePod)
+			if err != nil {
+				return errors.Wrapf(err, "error creating new pod: %s", source.Pod().GetName())
+			}
+
+			break
```
Comment (exdx) [inline pkg/controller/registry/reconciler/grpc.go]: I don't think we need to `break` here...

**Comment by exdx** [comment]
e2e now works locally (with skopeo installed)

File: pkg/api/apis/operators/v1alpha1/catalogsource_types.go [hunk 11]
```diff
@@ -54,6 +55,12 @@ type CatalogSourceSpec struct {
 	// +Optional
 	Image string `json:"image,omitempty"`
 
+	// Poll is used to determine the time interval between checks of the latest catalog source version.
+	// The catalog operator polls to see if a new version of the catalog source is available.
+	// If available, the latest image is pulled and gRPC traffic is directed to the latest catalog source.
+	// +Optional
+	Poll Poll `json:"poll,omitempty"`
```
[the remark this answers is not in the document]
Comment (exdx) [reply pkg/api/apis/operators/v1alpha1/catalogsource_types.go]: ok sounds good

**Comment by ecordell** [comment]
> pkg/api/apis/operators/v1alpha1/types_test.go:155:33: cannot use "k8s.io/apimachinery/pkg/apis/meta/v1".Time literal (type "k8s.io/apimachinery/pkg/apis/meta/v1".Time) as type *"k8s.io/apimachinery/pkg/apis/meta/v1".Time in field value

**Comment by exdx** [comment]
Hold on OLM CI image being updated. 
/hold

**Comment by exdx** [comment]
PR can go in once CI passes. TODO

- Refactor spec to use union fields per @njhale's suggestions for flexibility in adding new polling sources. Will require updating the feature and test unit/e2e code with the new spec. 
- Remove dependency on external quay registry and remove robot credentials: use openshift CI registry instead. catalog image will need to be copied there at the start of the test. 
- Document feature usage

**Comment by ecordell** [comment]
/lgtm

as discussed, we'll have follow up PRs for:

- cleaning up the logging
- removing quay creds
- updating the API definition in response to comments here

**Pull request merged**